### PR TITLE
Bug 1547062 - Experiment branch generation for trailhead

### PR DESCRIPTION
**Pull request by k88hudson**
Still need to write some tests/comments for this, but this has most of what we need

### CODE REVIEW

File: lib/ASRouter.jsm [hunk 1]
```diff
@@ -46,6 +81,50 @@ const MAX_MESSAGE_LIFETIME_CAP = 100;
...
+  let triplet;
+
+  // We can use a pref to override and bypass the experiment process
+  const overrideValue = Services.prefs.getStringPref(TRAILHEAD_OVERRIDE_PREF, "");
+  if (overrideValue) {
+    [interupt, triplet] = overrideValue.split("-");
+    return {experiment, interupt, triplet: triplet || ""};
+  }
+
+  const locale = Services.locale.appLocaleAsLangTag && Services.locale.appLocaleAsLangTag.substr(0, 2);
```
Comment (Mardak) [inline lib/ASRouter.jsm]: The locale code can be 3 letters although in our usage, there's none that are enX-, frX- or deX- anyway. Also `appLocaleAsLangTag` is guaranteed to be a string (so even if it's somehow empty string, substr will work)

File: lib/ASRouter.jsm [hunk 2]
```diff
@@ -28,7 +29,41 @@ ChromeUtils.defineModuleGetter(this, "QueryCache",
...
+ChromeUtils.defineModuleGetter(this, "TelemetryEnvironment",
+  "resource://gre/modules/TelemetryEnvironment.jsm");
+ChromeUtils.defineModuleGetter(this, "ClientEnvironment",
+  "resource://normandy/lib/ClientEnvironment.jsm");
+ChromeUtils.defineModuleGetter(this, "Sampling",
+  "resource://gre/modules/components-utils/Sampling.jsm");
+
+const TRAILHEAD_OVERRIDE_PREF = "trailhead.firstrun.branches";
+const TRAILHEAD_BRANCHES = {
+  interupts: [
```
Comment (Mardak) [inline lib/ASRouter.jsm]: I believe it's spelled "interrupts" (double "r" even in canadian 😉 )

File: lib/ASRouter.jsm [hunk 3]
```diff
@@ -28,7 +29,41 @@ ChromeUtils.defineModuleGetter(this, "QueryCache",
...
+    ["control"],
```
Comment (Mardak) [inline lib/ASRouter.jsm]: Did we want this to be "oldfxa" ?

File: lib/ASRouter.jsm [hunk 4]
```diff
@@ -618,13 +702,37 @@ class _ASRouter {
     }
   }
 
+  async setupTrailhead() {
+    if (this.state.trailheadInitialized) {
+      return;
+    }
+
+    const {experiment, interupt, triplet} = await generateTrailheadBranches();
+    await this.setState({trailheadInitialized: true, trailheadInterupt: interupt, trialheadTriplet: triplet});
```
Comment (Mardak) [inline lib/ASRouter.jsm]: Typo! `trailheadTriplet` (instead of `trial…`)

File: lib/ASRouter.jsm [hunk 5]
```diff
@@ -46,6 +81,50 @@ const MAX_MESSAGE_LIFETIME_CAP = 100;
 const LOCAL_MESSAGE_PROVIDERS = {OnboardingMessageProvider, CFRMessageProvider, SnippetsTestMessageProvider};
 const STARTPAGE_VERSION = "6";
 
+/**
+ * chooseBranch - choose from a list of optional branches and ratios
+ * @param branches {Array<[any, number?]>} A list of branches, where branch[0] is any item and branch[1] is the ratio
+ * @param seed {string} A unique seed for the randomizer
+ */
+async function chooseBranch(branches, seed) {
+  const ratios = branches.map(([item, ratio]) => ((typeof ratio !== "undefined") ? ratio : 1));
```
Comment (Mardak) [inline lib/ASRouter.jsm]: I suppose ratio could be `0` so we shouldn't just `ratio || 1` ?

File: lib/ASRouter.jsm [hunk 6]
```diff
@@ -46,6 +81,50 @@ const MAX_MESSAGE_LIFETIME_CAP = 100;
...
+  return branches[await Sampling.ratioSample(seed, ratios)][0];
```
Comment (Mardak) [inline lib/ASRouter.jsm]: A `@return` comment above or here to be clear we're returning the "any" portion of the branch.

File: lib/ASRouter.jsm (hunk 1 above)
Comment (k88hudson) [reply lib/ASRouter.jsm]: I'm updating this to be the full locale, since we don't need sub-buckets for locales

File: lib/ASRouter.jsm [hunk 7]
```diff
@@ -97,46 +91,46 @@ async function chooseBranch(branches, seed) {
...
   const overrideValue = Services.prefs.getStringPref(TRAILHEAD_CONFIG.OVERRIDE_PREF, "");
   if (overrideValue) {
-    [interupt, triplet] = overrideValue.split("-");
-    return {experiment, interupt, triplet: triplet || ""};
+    [interrupt, triplet] = overrideValue.split("-");
+    return {experiment, interrupt, triplet: triplet || ""};
   }
 
-  const locale = Services.locale.appLocaleAsLangTag && Services.locale.appLocaleAsLangTag.substr(0, 2);
+  const locale = Services.locale.appLocaleAsLangTag;
```
Comment (Mardak) [inline lib/ASRouter.jsm]: To be clear when we update the config.. This is what I see with mac de build:
```
Services.locale.appLocaleAsLangTag
"de"
```

File: lib/ASRouter.jsm [hunk 8]
```diff
@@ -55,7 +56,7 @@ const TRAILHEAD_CONFIG = {
   },
   LOCALES: ["en-US", "en-GB", "en-CA", "de", "de-DE", "fr", "fr-FR"],
   EXPERIMENT_RATIOS: [
-    ["", 98],
+    ["", 1],
```
Comment (Mardak) [inline lib/ASRouter.jsm]: this might be good to have for QA anyway

File: lib/ASRouter.jsm [hunk 9]
```diff
@@ -565,6 +566,8 @@ class _ASRouter {
     ASRouterPreferences.init();
     ASRouterPreferences.addListener(this.onPrefChange);
 
+    await this.setupTrailhead();
```
Comment (Mardak) [inline lib/ASRouter.jsm]: Probably a comment to keep updating Telemetry

**Pull request merged**